### PR TITLE
Add support for font-specific formats

### PR DESCRIPTION
**Pull request by Crissov**
- Text: ← filenames: FONTLOG http://scripts.sil.org/cms/scripts/page.php?item_id=OFL-FAQ_web#43cecb44
- OpenType: ← extensions: .fea https://www.adobe.com/devnet/opentype/afdko/topic_feature_file_syntax.html
- Spline Font: ← extensions: .sfd http://fontforge.github.io/en-US/documentation/developers/sfdformat/

### CODE REVIEW

File: lib/linguist/languages.yml [hunk 1]
```diff
@@ -2668,6 +2669,16 @@ OpenSCAD:
   tm_scope: none
   ace_mode: scad
 
+OpenType Font Feature:
+  type: programming
```
Comment (Alhadis) [inline lib/linguist/languages.yml]: If this PR is accepted, this should be `data`, not `programming`. Same with Spline Font Data.

Comment (Crissov) [inline lib/linguist/languages.yml]: I’ll agree that `data` is appropriate for `.sfd`, but not so much for `.fea`

Comment (Alhadis) [inline lib/linguist/languages.yml]: What's convinced you that OpenType features are a "programming language"...?

Comment (Crissov) [inline lib/linguist/languages.yml]: Well, maybe it’s `markup` (like Postscript and CSS) but OT features are used to implement substitution algorithms etc. That sounds like programming to me. It’s certainly not just `data` like an image or config file.

Comment (Alhadis) [inline lib/linguist/languages.yml]: PostScript? Uh, no, PostScript *is* a fully Turing-complete programming language... CSS, yes, that's markup.

I'm well aware of what OpenType features are, but I'd argue they're no closer to programming than SVG source is. They describe coordinates, geometry and hinting for the rendering environment...

Comment (Crissov) [inline lib/linguist/languages.yml]: I know that PS is TC and so is TeX, but like CSS they both have `type: markup` in `languages.yml`. CSS is a declarative language (and doesn’t mark up text)), but so are functional programming languages …

Comment (Alhadis) [inline lib/linguist/languages.yml]: The hell. I didn't even notice PostScript was classed as markup, and I was even editing that part the other night.

Right, time to amend.

**Comment by Alhadis** [comment]
You'll need to fix the order of file extensions. From the [build's feedback](https://travis-ci.org/github/linguist/jobs/149776048#L1159):

```
TestPedantic#test_nonprimary_extensions_are_sorted [/home/travis/build/github/linguist/test/test_pedantic.rb:15]:
.mkdown should come after .md.txt
```

**Comment by pchaigno** [comment]
@Crissov Thanks for the pull request! Please see the [guidelines to add support for a new extension to Linguist](https://github.com/github/linguist/blob/master/CONTRIBUTING.md#adding-an-extension-to-a-language) to complete the additional steps required.

**Comment by Alhadis** [comment]
@Crissov You can use [this package](https://github.com/Alhadis/language-fontforge) to provide syntax highlighting for OpenType feature files and spline-font databases. The relevant scopes are:

- **OpenType feature file:** `source.opentype`
- **SFD:** `text.sfd`

Also, it appears the "proper" name for the former is **"OpenType feature file"**, not "OpenType font feature". That's going by how [the spec refers to the format](http://www.adobe.com/devnet/opentype/afdko/topic_feature_file_syntax.html).

You'll also need to change OpenType's language type to `data`, as previous suggested. It's not markup.

File: lib/linguist/languages.yml [hunk 2]
```diff
@@ -2668,6 +2668,18 @@ OpenSCAD:
...
 
+Opentype feature:
+  type: data
+  aliases:
+  - AFDKO
+  extensions:
+  - .fea
+  #- .otl (rare)
+  #interpreters:
+  #- makeotf
```
Comment (Crissov) [inline lib/linguist/languages.yml]: I’m not sure I correctly understand the purpose of `interpreters`

**Comment by Alhadis** [comment]
Sorry about the long wait! Didn't mean to take this long to get back to you. 😞 

Alright, great news: all three additions check out.

### [`.sfd`](https://github.com/Alhadis/Linguist-Silo/tree/sfd#sfd)
Bit surprised by the abundance of these files; I hadn't expected FontForge's `.sfd` format was so well-distributed, heh. [11,018 files](https://github.com/search?q=extension%3Asfd+NOT+nothack&type=Code) on GitHub, hosted among [880 users](https://github.com/Alhadis/Linguist-Silo/blob/sfd/distribution.md#users) among [1,042 unique repositories](https://github.com/Alhadis/Linguist-Silo/blob/sfd/distribution.md#repositories). Definitely worth inclusion.

Each of the `.sfd` files [harvested from search results](https://github.com/Alhadis/Linguist-Silo/tree/sfd/files) turned out to be a Spline Font Database, so I don't predict there'll be any files misclassified here.

### [`.fea`](https://github.com/Alhadis/Linguist-Silo/tree/fea#fea)
These are slightly less distributed than `.sfd` files, but still too abundant to ignore: [14,236](https://github.com/search?q=extension%3Afea+NOT+nothack&type=Code) indexed results, and among the [2,695 files](https://github.com/Alhadis/Linguist-Silo/blob/fea/downloaded.md) I downloaded, half of which were duplicates, but distributed between 278 users among 404 unique repositories.

### `FONTLOG`
Can't really go wrong here... [17,730 results](https://github.com/search?q=filename%3AFONTLOG+NOT+nothack&type=Code), each with consistently [similar content and structure](https://github.com/markatango/langModel/blob/99030381df0fdc6cb4152e4bdcf5e9c23cfdcc58/Text_prediction/libraries/frameworks/landslide/themes/machinalis/fonts/FONTLOG).

---------------------

Now, you'll need to fix a few things before we can merge this:

* Pull/merge the latest changes from `master` into your branch
* Rerun `script/set-language-ids --update` to fix missing/conflicting language IDs
* Tweak each entry's name in `languages.yml`

/cc @pchaigno, @larsbrinkhoff

File: lib/linguist/languages.yml [hunk 3]
```diff
@@ -2668,6 +2668,15 @@ OpenSCAD:
   tm_scope: none
   ace_mode: scad
 
+Opentype feature:
```
Comment (Alhadis) [inline lib/linguist/languages.yml]: This needs to be changed to `OpenType Feature File` (case-sensitive, please).

File: lib/linguist/languages.yml [hunk 4]
```diff
@@ -3635,6 +3644,13 @@ SourcePawn:
   tm_scope: source.sp
   ace_mode: text
 
+Spline Font Data:
```
Comment (Alhadis) [inline lib/linguist/languages.yml]: The [canonical name](https://fontforge.github.io/sfdformat.html) for the format is `Spline Font Database`, which is what this needs to be. Again, please make sure it's capitalised correctly (the way I've written it here).

**Comment by Alhadis** [comment]
@Crissov Uhm, you've... kind of committed the conflicts you were supposed to fix. The diff for `languages.yml` is too large to open on GitHub, so I can't comment on the conflicted areas directly.

The affected regions are lines [2956-2969](https://github.com/Crissov/linguist/blob/18e4256b828c4186fec806319cbf8b76f0d2c79b/lib/linguist/languages.yml#L2956-L2969) and [4041-4052](https://github.com/Crissov/linguist/blob/18e4256b828c4186fec806319cbf8b76f0d2c79b/lib/linguist/languages.yml#L4041-L4052). Just erase the the blank lines and `<<<<====>>>>` junk, and rerun `script/set-language-ids --update`.

**Comment by Alhadis** [comment]
@Crissov Just paging you with a reminder we're still waiting for the conflicts to be fixed before we can finish our review. =)

**Comment by brandonblack** [comment]
@Crissov looks like this is near the finish line, but you have some additional conflicts to resolve.

**Comment by Alhadis** [comment]
@Crissov I sense you may be having difficulties, so I [sent you a pull request](https://github.com/Crissov/linguist/pull/1) to resolve these conflicts.

**Comment by Alhadis** [comment]
Alright, there're two build failures, but only one of them we can fix (the [second failure](https://travis-ci.org/github/linguist/jobs/177226335#L1118) isn't your fault; it's currently a [known issue](https://github.com/github/linguist/pull/3337)).

This is the one we still have to fix:
```
  1) Failure:
TestLanguage#test_all_languages_have_grammars [/home/travis/build/github/linguist/test/test_language.rb:409]:
The following languages' scopes are not listed in grammars.yml. Please add grammars for all new languages.
If no grammar exists for a language, mark the language with `tm_scope: none` in lib/linguist/languages.yml.
OpenType Feature File source.opentype
Spline Font Database  text.sfd
```
I admit the error here was probably mine, since I hastily deleted my fork before seeing what the build result was. Silly me.

@Crissov Try these steps:
1. Clone your fork to a **new directory** on your system.
2. Checkout your branch `patch-2`
3. Run `script/bootstrap` to install the necessary submodules
4. Run the following:

```sh
perl -pi -e 's|script\K (?=list-grammars)|/|;' script/add-grammar
script/add-grammar Alhadis/language-fontforge
git add --all && git commit -m "Update submodules and fix grammar-list script"
```

Then push to your fork as you would normally.

Using `script/add-grammar` is the simplest fix here, but the current state of the script isn't working due an error in 43fa563b. I've added an extra step that fixes it in order to allow the script to run.

Probably not a clean way to do it, but hey.

**Comment by Alhadis** [comment]
@Crissov I've sent you a [second pull request](https://github.com/Crissov/linguist/pull/2) in the hopes of closing the book on this.

@pchaigno Tests were still failing locally for the Ant submodule (even after merging master into the fork's branch). Could this be related to `licensee` somehow?

**Comment by pchaigno** [comment]
@Alhadis You'll need to update dependencies: `bundle update`.

**Comment by Alhadis** [comment]
Ah, right. My understanding of what `bundle` even is is still really shaky.

Is it because I'm from the land of JavaScript, not Ruby? ;)

**Comment by pchaigno** [comment]
> Is it because I'm from the land of JavaScript, not Ruby? ;)

I'm not sure I get it myself, but then I'm not sure either of which land I'm from :p 

> Actually, serious question: shouldn't there be a version check or something to ensure dependencies are up to date? Because I had no idea what was going on.

We could increase the version of Licensee needed in [github-linguist.gemspec](https://github.com/github/linguist/blob/829eea01396e8e5747cd83f799d41196b2b9d898/github-linguist.gemspec). I'm a little hesitant to do that because Licensee 8.7.0 isn't actually needed for anything else than tests.

**Comment by Alhadis** [comment]
@Crissov Could you please merge the PR so we can get this wrapped up?

**Comment by Alhadis** [comment]
@brandonblack This looks merge-ready to me. :shipit: ?

Please also remember to squash-merge; the PR's commit history is extremely convoluted and would pollute the codebase's revision history.

**Comment by brandonblack** [comment]
Thanks @Crissov!

**Comment by Crissov** [comment]
It was really @Alhadis who did all the nasty work.

**Pull request merged**